app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala

```diff
@@ -0,0 +1,119 @@
+import play.api.data.Form
+import play.api.i18n.I18nSupport
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import uk.gov.hmrc.plasticpackagingtax.registration.connectors.{
+  IncorpIdConnector,
+  RegistrationConnector,
+  ServiceError
+}
+import uk.gov.hmrc.plasticpackagingtax.registration.controllers.actions.{
+  AuthAction,
+  FormAction,
+  SaveAndContinue
+}
+import uk.gov.hmrc.plasticpackagingtax.registration.forms.ConfirmAddress
+import uk.gov.hmrc.plasticpackagingtax.registration.forms.ConfirmAddress.form
+import uk.gov.hmrc.plasticpackagingtax.registration.models.genericregistration.IncorporationAddressDetails
+import uk.gov.hmrc.plasticpackagingtax.registration.models.registration.{Cacheable, Registration}
+import uk.gov.hmrc.plasticpackagingtax.registration.models.request.{JourneyAction, JourneyRequest}
+import uk.gov.hmrc.plasticpackagingtax.registration.views.html.confirm_address
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendController
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class ContactDetailsConfirmAddressController @Inject() (
+  authenticate: AuthAction,
+  journeyAction: JourneyAction,
+  override val registrationConnector: RegistrationConnector,
+  incorpIdConnector: IncorpIdConnector,
+  mcc: MessagesControllerComponents,
+  page: confirm_address
+)(implicit ec: ExecutionContext)
+    extends FrontendController(mcc) with Cacheable with I18nSupport {
+
+  def displayPage(): Action[AnyContent] =
+    (authenticate andThen journeyAction).async { implicit request =>
+      request.registration.incorpJourneyId match {
+        case Some(journeyId) =>
+          for {
+            incorporationDetails <- incorpIdConnector.getDetails(journeyId)
+          } yield Ok(
+            page(form().fill(
+                   ConfirmAddress(request.registration.primaryContactDetails.useRegisteredAddress)
+                 ),
+                 incorporationDetails.companyAddress
+            )
+          )
+        case _ => Future(Redirect(routes.RegistrationController.displayPage()))
+      }
+    }
+
+  def submit(): Action[AnyContent] =
+    (authenticate andThen journeyAction).async { implicit request =>
+      request.registration.incorpJourneyId match {
+        case Some(journeyId) =>
+          incorpIdConnector.getDetails(journeyId).flatMap { incorporationDetails =>
+            form()
+              .bindFromRequest()
+              .fold(
+                (formWithErrors: Form[ConfirmAddress]) =>
+                  Future(BadRequest(page(formWithErrors, incorporationDetails.companyAddress))),
+                confirmAddress =>
+                  updateRegistration(confirmAddress, incorporationDetails.companyAddress).map {
+                    case Right(_) =>
+                      FormAction.bindFromRequest match {
+                        case SaveAndContinue =>
+                          if (confirmAddress.useRegisteredAddress.getOrElse(false))
```

Comment:
I think we could use some methods of Option here:
`confirmAddress.useRegisteredAddress
                            .map( _ => Redirect(routes.ContactDetailsCheckAnswersController.displayPage()))
                            .getOrElse(Redirect(routes.ContactDetailsAddressController.displayPage()))`
what do you think? (feel free to discard)

Reply:
No because we can only redirect to `routes.ContactDetailsCheckAnswersController.displayPage()` if `true`

Reply:
aa cool, yeap I can see that now